make/common/NativeCompilation.gmk

```diff
@@ -1130,9 +1137,56 @@ define SetupNativeCompilationBody
     endif
   endif
 
+  $1_LD_OBJ_ARG := $$($1_ALL_OBJS)
+
+  # If there are many object files, use an @-file...
+  ifneq ($$(word 17, $$($1_ALL_OBJS)), )
+    $1_OBJ_FILE_LIST := $$($1_OBJECT_DIR)/_$1_objectfilenames.txt
+    ifneq ($(COMPILER_COMMAND_FILE_FLAG), )
+      $1_LD_OBJ_ARG := $(COMPILER_COMMAND_FILE_FLAG)$$($1_OBJ_FILE_LIST)
+    else
+      # ...except for toolchains which don't support them.
+      $1_LD_OBJ_ARG := `cat $$($1_OBJ_FILE_LIST)`
+    endif
+
+    # If we are building static library, 'AR' on macosx may not support @-file.
+    ifeq ($$($1_TYPE), STATIC_LIBRARY)
+      ifeq ($(call isTargetOs, macosx), true)
+        $1_LD_OBJ_ARG := `cat $$($1_OBJ_FILE_LIST)`
+      endif
+    endif
+  endif
+
+  # Unfortunately the @-file trick does not work reliably when using clang.
+  # Clang does not propagate the @-file parameter to the ld sub process, but
+  # instead puts the full content on the command line. At least the llvm ld
+  # does not even support an @-file.
+  #
+  # When linking a large amount of object files, we risk hitting the limit
+  # of the command line length even on posix systems if the path length of
+  # the output dir is very long due to our use of absolute paths. To
+  # mitigate this, use paths relative to the output dir when linking over
+  # 500 files with clang and the output dir path is deep.
+  ifneq ($$(word 500, $$($1_ALL_OBJS)), )
+    ifeq ($$(TOOLCHAIN_TYPE), clang)
+      # There is no strlen function in make, but checking path depth is a
+      # reasonable approximation.
+      ifneq ($$(word 10, $$(subst /, ,$$(OUTPUTDIR))), )
+        $1_LINK_OBJS_RELATIVE := true
+        $1_ALL_OBJS_RELATIVE := $$(patsubst $$(OUTPUTDIR)/%, %, $$($1_ALL_OBJS))
+      endif
+    endif
+  endif
+
   ifeq ($$($1_TYPE), STATIC_LIBRARY)
     $1_VARDEPS := $$($1_AR) $$(ARFLAGS) $$($1_ARFLAGS) $$($1_LIBS) \
         $$($1_EXTRA_LIBS)
+    # Include partial linking when building the static library with clang on linux.
+    ifeq ($(call isTargetOs, linux), true)
+      ifneq ($(findstring $(TOOLCHAIN_TYPE), clang), )
```

Comment:
This combination of conditions is repeated 3 times. Maybe we could assign the result to a variable (e.g. `$1_ENABLE_PARTIAL_LINKING`) and check for that instead?

Reply:
> This combination of conditions is repeated 3 times. Maybe we could assign the result to a variable (e.g. `$1_ENABLE_PARTIAL_LINKING`) and check for that instead?

Done, thanks.